x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/LifecycleExecutionState.java

```diff
@@ -0,0 +1,253 @@
+     * @param indexMetaData The metadata of the index to retrieve the execution
+     *                      state from.
+     * @return The execution state of that index.
+     */
+    public static LifecycleExecutionState fromIndexMetadata(IndexMetaData indexMetaData) {
+        Map<String, String> customData = indexMetaData.getCustomData(ILM_CUSTOM_METADATA_KEY);
+        customData = customData == null ? new HashMap<>() : customData;
+        return fromCustomMetadata(customData);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static Builder builder(LifecycleExecutionState state) {
+        return new Builder()
+            .setPhase(state.phase)
+            .setAction(state.action)
+            .setStep(state.step)
+            .setFailedStep(state.failedStep)
+            .setStepInfo(state.stepInfo)
+            .setPhaseDefinition(state.phaseDefinition)
+            .setIndexCreationDate(state.indexCreationDate)
+            .setPhaseTime(state.phaseTime)
+            .setActionTime(state.actionTime)
+            .setStepTime(state.stepTime);
+    }
+
+    static LifecycleExecutionState fromCustomMetadata(Map<String, String> customData) {
+        return new Builder()
+            .setPhase(customData.getOrDefault(PHASE, ""))
+            .setAction(customData.getOrDefault(ACTION, ""))
+            .setStep(customData.getOrDefault(STEP, ""))
+            .setFailedStep(customData.getOrDefault(FAILED_STEP, ""))
+            .setStepInfo(customData.getOrDefault(STEP_INFO, ""))
+            .setPhaseDefinition(customData.getOrDefault(PHASE_DEFINITION, ""))
+            .setIndexCreationDate(Long.parseLong(customData.getOrDefault(INDEX_CREATION_DATE, "-1")))
+            .setPhaseTime(Long.parseLong(customData.getOrDefault(PHASE_TIME, "-1")))
+            .setActionTime(Long.parseLong(customData.getOrDefault(ACTION_TIME, "-1")))
+            .setStepTime(Long.parseLong(customData.getOrDefault(STEP_TIME, "-1")))
+            .build();
+    }
+
+    /**
+     * Converts this object to an immutable map representation for use with
+     * {@link IndexMetaData.Builder#putCustom(String, Map)}.
+     * @return An immutable Map representation of this execution state.
+     */
+    public Map<String, String> asMap() {
+        Map<String, String> result = new HashMap<>();
+        result.put(PHASE, phase);
+        result.put(ACTION, action);
+        result.put(STEP, step);
+        result.put(FAILED_STEP, failedStep);
+        result.put(STEP_INFO, stepInfo);
+        result.put(INDEX_CREATION_DATE, String.valueOf(indexCreationDate));
+        result.put(PHASE_TIME, String.valueOf(phaseTime));
+        result.put(ACTION_TIME, String.valueOf(actionTime));
+        result.put(STEP_TIME, String.valueOf(stepTime));
+        result.put(PHASE_DEFINITION, String.valueOf(phaseDefinition));
+        return Collections.unmodifiableMap(result);
+    }
+
+    public String getPhase() {
```

Comment:
This may be my (semi controversial) opinion, but I think we could ditch the getters and make the vars public since this class does nothing with the values; what do you think?